escrow/src/main.sw

```diff
@@ -1,257 +1,346 @@
+            // Not needed as long as the entire contract handles state correctly but leaving it in
+            // for conceptual closure at the slight expense of users
+            storage.arbiter_proposal.insert(identifier, Option::None);
+        }
+
+        log(ResolvedDisputeEvent {
+            identifier, user
+        });
     }
 
-    /// Returns the amount of the specified asset in this contract
-    #[storage(read)]fn get_balance() -> u64 {
-        this_balance(storage.asset)
+    #[storage(read, write)]fn return_deposit(identifier: u64) {
+        // The assertions ensure that only the seller can return the deposit as long as the escrow
+        // contains a deposit and the escrow has not been completed
+
+        let mut escrow = storage.escrows.get(identifier);
+
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+        require(msg_sender().unwrap() == escrow.seller.address, UserError::Unauthorized);
+        require(escrow.buyer.asset.is_some() && 0 < escrow.buyer.deposited_amount, StateError::CannotTransferBeforeDesposit);
+
+        escrow.state = State::Completed;
+        storage.escrows.insert(identifier, escrow);
+
+        transfer(escrow.buyer.deposited_amount, escrow.buyer.asset.unwrap(), escrow.buyer.address);
+        transfer(escrow.arbiter.fee_amount, escrow.arbiter.asset, escrow.seller.address);
+
+        // If there is a previous proposal then we must transfer those funds back to the seller
+        let proposal = storage.arbiter_proposal.get(identifier);
+        if proposal.is_some() && 0 < proposal.unwrap().fee_amount {
+            transfer(proposal.unwrap().fee_amount, proposal.unwrap().asset, escrow.seller.address);
+            // Not needed as long as the entire contract handles state correctly but leaving it in
+            // for conceptual closure at the slight expense of users
+            storage.arbiter_proposal.insert(identifier, Option::None);
+        }
+
+        log(ReturnedDepositEvent {
+            identifier
+        });
     }
 
-    /// Returns data regarding the state of a user i.e. whether they have (deposited, approved)
-    ///
-    /// # Panics
-    ///
-    /// The function will panic when
-    /// - The constructor has not been called to initialize
-    /// - The user is not an authorized user that has been set in the constructor
-    #[storage(read)]fn get_user_data(user: Address) -> (bool, bool) {
-        // require(storage.state != State::Void, Error::StateNotInitialized);
-        require(storage.state != 0, Error::StateNotInitialized);
-        require(user == storage.buyer.address || user == storage.seller.address, Error::UnauthorizedUser);
-
-        if user == storage.buyer.address {
-            (storage.buyer.deposited, storage.buyer.approved)
-        } else {
-            (storage.seller.deposited, storage.seller.approved)
+    #[storage(read, write)]fn take_payment(identifier: u64) {
```

Comment:
My understanding is that the buyer receives the selling asset off chain, hence the need for an arbiter

Reply:
The way the process works is the buyer deposits an asset into the escrow and then they wait for something to happen off-chain. Once that occurs they can use the `transfer_to_seller()` to send their deposit to the seller. If for whatever reason they do not do that and the deadline is surpassed then the seller can `take_payment()` instead. If the buyer does not want to allow that, because the trade did not occur as expected, they can `dispute()` at which point an arbiter must come in to resolve OR if the arbiter does not resolve in time the seller can `return_deposit()` or the buyer can still `transfer_to_seller()`

Reply:
Ah I see, the off-chain portion went over my head. This makes a lot more sense.